pkg/serviceaccount/jwt.go

```diff
@@ -53,43 +56,143 @@ type TokenGenerator interface {
+	// Derive the KeyID from the public key, in a non-reversible way.  The
+	// KeyID helps OIDC verifiers cope with key rotations.  Making the
+	// derivation non-reversible makes it impossible for someone to
+	// accidentally obtain the real key from the key ID and use it for token
+	// validation.
+
+	publicKeyDERBytes, err := x509.MarshalPKIXPublicKey(&keyPair.PublicKey)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to serialize public key to DER format")
+	}
+
+	hasher := crypto.SHA256.New()
+	hasher.Write(publicKeyDERBytes)
+	publicKeyDERHash := hasher.Sum(nil)
+
+	keyID := base64.RawURLEncoding.EncodeToString(publicKeyDERHash)
+
+	// Wrap the RSA keypair in a JOSE JWK with the designated key ID.
+	privateJWK := &jose.JSONWebKey{
+		Algorithm: string(jose.RS256),
+		Key:       keyPair,
+		KeyID:     keyID,
+		Use:       "sig",
+	}
+
+	signer, err := jose.NewSigner(
+		jose.SigningKey{
+			Algorithm: jose.RS256,
+			Key:       privateJWK,
+		},
+		nil,
+	)
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create signer")
+	}
+
+	return signer, nil
+}
+
+func signerFromECDSAPrivateKey(keyPair *ecdsa.PrivateKey) (jose.Signer, error) {
+	var alg jose.SignatureAlgorithm
+	switch keyPair.Curve {
+	case elliptic.P256():
+		alg = jose.ES256
+	case elliptic.P384():
+		alg = jose.ES384
+	case elliptic.P521():
+		alg = jose.ES512
+	default:
+		return nil, fmt.Errorf("unknown private key curve, must be 256, 384, or 521")
+	}
+
+	// Derive the KeyID from the public key, in a non-reversible way.  The
+	// KeyID helps OIDC verifiers cope with key rotations.  Making the
+	// derivation non-reversible makes it impossible for someone to
+	// accidentally obtain the real key from the key ID and use it for token
+	// validation.
+
+	publicKeyDERBytes, err := x509.MarshalPKIXPublicKey(&keyPair.PublicKey)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to serialize public key to DER format")
+	}
+
+	hasher := crypto.SHA256.New()
+	hasher.Write(publicKeyDERBytes)
+	publicKeyDERHash := hasher.Sum(nil)
+
+	keyID := base64.RawURLEncoding.EncodeToString(publicKeyDERHash)
+
+	// Wrap the ECDSA keypair in a JOSE JWK with the designated key ID.
+	privateJWK := &jose.JSONWebKey{
+		Algorithm: string(alg),
+		Key:       keyPair,
+		KeyID:     keyID,
+		Use:       "sig",
+	}
+
 	signer, err := jose.NewSigner(
 		jose.SigningKey{
 			Algorithm: alg,
-			Key:       privateKey,
+			Key:       privateJWK,
 		},
 		nil,
 	)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "failed to create signer")
 	}
-	return &jwtTokenGenerator{
-		iss:    iss,
-		signer: signer,
-	}, nil
+
+	return signer, nil
+}
+
+func signerFromOpaqueSigner(opaqueSigner jose.OpaqueSigner) (jose.Signer, error) {
+	alg := jose.SignatureAlgorithm(opaqueSigner.Public().Algorithm)
+
+	signer, err := jose.NewSigner(
+		jose.SigningKey{
+			Algorithm: alg,
+			Key:       opaqueSigner,
```

Comment:
Opaque signers should also forward key ids:
```
KeyID: opaqueSigner.Public().KeyID
```
Key rotation is going to be tricky though.

Reply:
Done

Reply:
Actually, I spoke too soon --- `SigningKey` doesn't have a `KeyID` field.  Is it safe to change the Key field from `opaqueSigner` to a `JSONWebKey` that wraps `opaqueSigner`?

```
signer, err := jose.NewSigner(
    jose.SigningKey{
        Algorithm: alg,
        Key: &jose.JSONWebKey{
            Algorithm: alg,
            Key: opaqueSigner,
            KeyID: opaqueSigner.Public().KeyID,
            Use: "sig",
        },
    },
    nil
)